lib/RestructureCFG/InlineDivergentScopesPass.cpp

```diff
@@ -0,0 +1,396 @@
+      // promote a case already pointing to `PlaceHolderTarget`, this would, in
+      // turn, cause the `default` case to not be simplified ever.
+      for (auto CaseIt = Switch->case_begin(); CaseIt != Switch->case_end();
+           ++CaseIt) {
+        if (CaseIt->getCaseSuccessor() != PlaceHolderTarget) {
+          Switch->setDefaultDest(CaseIt->getCaseSuccessor());
+          Switch->removeCase(CaseIt);
+          break;
+        }
+      }
+    }
+
+    // Handle the simplification when `PlaceHolderTarget` is part the standard
+    // `case`s
+    for (auto CaseIt = Switch->case_begin(); CaseIt != Switch->case_end();) {
+      if (CaseIt->getCaseSuccessor() == PlaceHolderTarget) {
+
+        // We do not want to have a situation where the `PlaceHolderTarget` is
+        // both the `default` successor of a `switch` and one of its standard
+        // case
+        CaseIt = Switch->removeCase(CaseIt);
+      } else {
+        ++CaseIt;
+      }
+    }
+
+    // It should never be the case that we end up with a `switch` having only
+    // `PlaceHolderTarget` as its successor
+    if (Switch->getNumCases() == 0
+        and Switch->getDefaultDest() == PlaceHolderTarget) {
+      revng_abort();
+    }
+  }
+}
+
+/// We define a `Descriptor` for the information needed to identify a
+/// `Divergence`. Specifically, the divergence is composed by the conditional
+/// where divergence originates, a set of divergent successors, and by the exit
+/// node which is divergent wrt. those divergent successors.
+struct DivergenceDescriptor {
+  BasicBlock *Conditional;
+  SmallSet<BasicBlock *, 4> DivergentSuccessors;
+  BasicBlock *Exit;
+};
+
+/// Helper function that tries to identify a `DivergenceDescriptor`
+static std::optional<DivergenceDescriptor>
+electDivergence(BasicBlock *Candidate,
+                BasicBlock *Exit,
+                DomTreeOnView<BasicBlock, Scope> &DomTree) {
+
+  // Check if it is the conditional making it a divergent node
+  if (isConditional(Candidate) and DomTree.dominates(Candidate, Exit)) {
+    using ScopeBlock = Scope<BasicBlock *>;
+
+    SetVector<BasicBlock *> Successors;
+    for (BasicBlock *Successor : children<ScopeBlock>(Candidate)) {
+      Successors.insert(Successor);
+    }
+
+    MapVector<BasicBlock *, SmallSet<BasicBlock *, 4>> ReachablesFromSuccessor;
+    size_t SuccessorsSize = std::distance(Successors.begin(), Successors.end());
+
+    for (auto *Successor : Successors) {
+      SmallSet<BasicBlock *, 4> ReachableExits;
+
+      // Explore all the exits that are reachable from each Successor.
+      // For having a divergent exit, we need to find a set of
+      // successors that reach only the exit under analysis
+      for (auto *DFSNode : depth_first(Scope<BasicBlock *>(Successor))) {
+        if (isExit(DFSNode)) {
+          ReachableExits.insert(DFSNode);
+        }
+      }
+
+      ReachablesFromSuccessor[Successor] = ReachableExits;
+    }
```

Comment:
Nice that you're using a `MapVector`.
But this is yet another case where you can outline this computation in a separate standalone free `static` function.

Also, `ReachableExits` isn't necessary.

Just:
```c++
      auto &Reachables = ReacheblesFromSuccessor[Successor];
      for (auto *DFSNode : depth_first(Scope<BasicBlock *>(Successor))) {
        if (isExit(DFSNode)) {
          Reachables.insert(DFSNode);
        }
      }
```

Reply:
Oh right, I factored this out (I also took out the `Successors` collection slightly above, since I only care about the `size` here https://github.com/revng/revng/blob/9208e1d0d7ecf56a86e3f44debb589c8eded149c/lib/RestructureCFG/InlineDivergentScopesPass.cpp#L164